goalie/field.py

```
@@ -105,3 +96,46 @@ def __eq__(self, other):
 
     def __ne__(self, other):
         return not self.__eq__(other)
+
+    def get_element(self, mesh):
+        """
+        Given a mesh, return the finite element associated with the field.
+
+        :arg mesh: The mesh to determine the cell from.
+        :type mesh: :class:`~.firedrake.mesh.MeshGeometry`
+        :return: The finite element associated with the field.
+        :rtype: An appropriate subclass of :class:`~.FiniteElementBase`
+        """
+        if self.finite_element is not None:
+            assert self.finite_element.cell == mesh.coordinates.ufl_element().cell
+            return self.finite_element
+
+        if self.family is not None:
+            finite_element = make_scalar_element(
+                mesh,
+                self.family,
+                self.degree,
+                self.vfamily,
+                self.vdegree,
+                self.variant,
+            )
+
+            if self.vector:
+                finite_element = VectorElement(
+                    finite_element, dim=finite_element.cell.topological_dimension()
+                )
+            return finite_element
+
+        return FiniteElement("Real", ufl.interval, 0)
```

But what about if you specify e.g., `degree` but not `family`? This should raise an error but I think it'd get past the logic here? The `degree` that was passed would just be silently ignored and you'd get a Real space, which might be unexpected.

---

Yeah, this also wasn't clear to me. Apparently `make_scalar_element` only requires `mesh` and `family` to be specified. Everything else can be `None`. For example:
```
finite_element = make_scalar_element(
    UnitIntervalMesh(1),
    family="Real",
    degree=None,
    vfamily=None,
    vdegree=None,
    variant=None,
)
print(finite_element)

fspace = FunctionSpace(mesh, finite_element)
print(fspace)
```
prints
```
<R? on a interval>
WithGeometry(FunctionSpace(<firedrake.mesh.MeshTopology object at 0x7f7f2dcf3da0>, FiniteElement('Real', interval, None), name=None), <Mesh #0>)
```
So I'm not entirely sure how that works :) but I guess we shouldn't raise an error here if Firedrake doesn't raise it?

And if we pass `degree` but not `family`, as you commented, we get `ValueError: Unknown finite element 'None'.` raised by `make_scalar_element`.

So to be honest, I think the best solution is to avoid setting the default Real space. I think it's better to be explicit. So that's what I did in https://github.com/mesh-adaptation/goalie/pull/301/commits/650bf1273f11cee7814db5fd978af3f7176fb954 and https://github.com/mesh-adaptation/goalie/pull/301/commits/337d89affd807cde1fb687e81179bcc4eac76cb1, where I always require that either `finite_element` or `family` are specified. I think you said that you added this default behaviour just to make testing easier, but it doesn't seem to be such a big change. What do you think?

---

Works for me!